### PR TITLE
Implement Pattern Matching in TruffleRuby (GSoC)

### PR DESCRIPTION
**Pull request by razetime**
Work done for the Ruby GSoC program in 2022, aiming to complete pattern matching tasks in #2453.
**Proposal name:** Implementing Ruby 3.0 Pattern Matching in TruffleRuby

### CODE REVIEW

File: src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+        currentValueToMatch = expressionValue;
+
+        // TODO move the other cases to the visitor pattern
+        switch (patternNode.getNodeType()) {
+            case ARRAYPATTERNNODE:
+                return this.visitArrayPatternNode((ArrayPatternParseNode) patternNode);
+            case HASHPATTERNNODE:
+                return this.visitHashPatternNode((HashPatternParseNode) patternNode);
+            case FINDPATTERNNODE:
+                //                throw CompilerDirectives.shouldNotReachHere();
```
Comment (eregon) [inline src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java:300]: Let's remove this line

File: src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+            Source source,
+            ParserContext parserContext,
+            Node currentNode,
+            ParseNode data,
+            ListParseNode cases,
+            TranslatorEnvironment environment,
+            BodyTranslator bodyTranslator) {
+        super(language, source, parserContext, currentNode, environment);
+        this.data = data; // data to match on
+        this.cases = cases; // cases to check.
```
Comment (eregon) [inline src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java:79]: These comments would be better on the fields above, as javadoc comments (`/** foo */`)

File: src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+import org.truffleruby.parser.ast.TrueParseNode;
+
+import java.util.Arrays;
+
+public class PatternMatchingTranslator extends BaseTranslator {
+
+    ParseNode data;
+    ListParseNode cases;
+    TranslatorEnvironment environment;
+    BodyTranslator bodyTranslator;
```
Comment (eregon) [inline src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java:64]: These 4 fields should be `final`

File: src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+
+    //    public RubyNode translateArrayPatternNode(ArrayPatternParseNode node, ArrayParseNode data) {
+    //        // For now, we are assuming that only preArgs exist, and the pattern only consists of simple constant values.
+    //        final int size = node.minimumArgsNum();
+    //        ListParseNode pre = node.getPreArgs();
+    //        ParseNode[] ch = pre.children();
+    //        for (int i = 0; i < pre.size(); i++) {
+    //            if (ch[i] === data ) {}
+    //        }
+    //    }
```
Comment (eregon) [inline src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java:436]: Remove this commented code

**Review by eregon — COMMENTED** [review]
Also remove `src/main/java/org/truffleruby/parser/parser/y.output` and `pattern_matching.rb`

File: src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java (hunk 1 above)
Comment (razetime) [reply src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java]: removed

File: src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+    @Override
+    protected RubyNode defaultVisit(ParseNode node) {
+        final RubyContext context = RubyLanguage.getCurrentContext();
+        throw new RaiseException(
+                context,
+                context.getCoreExceptions().syntaxError(
+                        "not yet handled in pattern matching: " + node.toString() + " " + node.getPosition(),
+                        currentNode,
+                        node.getPosition().toSourceSection(source)));
+        //        throw new UnsupportedOperationException(node.toString() + " " + node.getPosition());
```
[the remark this answers is not in the document]
Comment (razetime) [reply src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java]: removed

File: src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java (hunk 2 above)
Comment (razetime) [reply src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java]: converted to javadoc comments

File: src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java (hunk 3 above)
Comment (razetime) [reply src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java]: made `final`

File: src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java (hunk 4 above)
Comment (razetime) [reply src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java]: removed

File: src/main/java/org/truffleruby/language/control/CheckIfPatternsMatchedNode.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+    @Child RubyNode inspected;
+
+    public CheckIfPatternsMatchedNode(RubyNode inspected) {
+        this.inspected = inspected;
+    }
+
+    @Override
+    public Object execute(VirtualFrame frame) {
+        String message = inspected.execute(frame).toString();
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/control/CheckIfPatternsMatchedNode.java:27]: See https://github.com/oracle/truffleruby/actions/runs/3507592176/jobs/5887990366
`org.truffleruby.language.control.CheckIfPatternsMatchedNode.execute(CheckIfPatternsMatchedNode.java:27) ; java.util.AbstractCollection.toString() ; java.lang.Object.toString()`
which means `CheckIfPatternsMatchedNode.java:27` calls toString but that can't be called in PE code.
It's best to use a ToJavaStringNode here to convert from Ruby String to Java String.

Comment (razetime) [reply src/main/java/org/truffleruby/language/control/CheckIfPatternsMatchedNode.java:27]: I am not entirely sure how to use  `ToJavaStringNode` with a `RubyString`. Will `public String getJavaString()` from `TStringUtils` work?

Comment (eregon) [reply src/main/java/org/truffleruby/language/control/CheckIfPatternsMatchedNode.java:27]: No, that's neverPartOfCompilation() so also not allowed in PE code.
Use `@Child  ToJavaStringNode  toJavaStringNode` + `toJavaStringNode.execute(inspected.execute(frame)`

Comment (razetime) [reply src/main/java/org/truffleruby/language/control/CheckIfPatternsMatchedNode.java:27]: can i do `inspected.execute(frame).asTruffleString()` or `asTruffleStringUncached()`?

Comment (razetime) [reply src/main/java/org/truffleruby/language/control/CheckIfPatternsMatchedNode.java:27]: this works:
```
        toJavaStringNode = TruffleString.ToJavaStringNode.create();
        String message = toJavaStringNode.execute(((RubyString) inspected.execute(frame)).asTruffleStringUncached());
        throw new RaiseException(getContext(), coreExceptions().noMatchingPatternError(message, this));
```

Comment (eregon) [reply src/main/java/org/truffleruby/language/control/CheckIfPatternsMatchedNode.java:27]: Anything uncached is not allowed in PE code (e.g., asTruffleStringUncached).
What does `inspected.execute(frame)` return? Isn't it already a Ruby String?

Comment (razetime) [reply src/main/java/org/truffleruby/language/control/CheckIfPatternsMatchedNode.java:27]: `inspected.execute(frame)` returns a `RubyString`, but `toJavaStringNode` expects a truffleString. I get a typeError regardless of a cast.

Comment (eregon) [reply src/main/java/org/truffleruby/language/control/CheckIfPatternsMatchedNode.java:27]: It should be `org.truffleruby.interop.ToJavaStringNode` and not `com.oracle.truffle.api.strings.TruffleString.ToJavaStringNode`

Comment (razetime) [reply src/main/java/org/truffleruby/language/control/CheckIfPatternsMatchedNode.java:27]: I have implemented the change. I hope it is correct.

File: src/main/java/org/truffleruby/language/control/CheckIfPatternsMatchedNode.java [hunk 7]
```diff
@@ -17,14 +18,16 @@
...
 
     public CheckIfPatternsMatchedNode(RubyNode inspected) {
         this.inspected = inspected;
     }
 
     @Override
     public Object execute(VirtualFrame frame) {
-        String message = inspected.execute(frame).toString();
+        toJavaStringNode = ToJavaStringNode.create(inspected);
+        String message = (String) toJavaStringNode.execute(frame);
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/control/CheckIfPatternsMatchedNode.java:30]: Use `String message = toJavaStringNode.executeToJavaString(inspected.execute(frame));` and then no need to cast

File: src/main/java/org/truffleruby/language/control/CheckIfPatternsMatchedNode.java [hunk 8]
```diff
@@ -17,14 +18,16 @@
...
+    @Child ToJavaStringNode toJavaStringNode;
 
     public CheckIfPatternsMatchedNode(RubyNode inspected) {
         this.inspected = inspected;
     }
 
     @Override
     public Object execute(VirtualFrame frame) {
-        String message = inspected.execute(frame).toString();
+        toJavaStringNode = ToJavaStringNode.create(inspected);
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/control/CheckIfPatternsMatchedNode.java:29]: We must not create nodes in PE code. So it should `toJavaStringNode = ToJavaStringNode.create();` in the constructor.

**Review by eregon — APPROVED** [review]
It took a very long time for me to review and integrate this, I am sorry.

OTOH we already reviewed the code during the GSoC so the PR was in a pretty good state already.
Mostly I `git filter-repo`'d away the big files that were temporarily committed to avoid increasing the repo size, and I did a bunch of cleanups.

**Pull request merged**